pkg/webhook/certs.go

```diff
@@ -0,0 +1,486 @@
+		return nil, nil, errors.Wrap(err, "encoding key")
+	}
+	return certBuf.Bytes(), keyBuf.Bytes(), nil
+}
+
+func lookaheadTime() time.Time {
+	return time.Now().Add(90 * 24 * time.Hour)
+}
+
+func validServerCert(caCert, cert, key []byte) bool {
+	valid, err := validCert(caCert, cert, key, DNSName, lookaheadTime())
+	if err != nil {
+		return false
+	}
+	return valid
+}
+
+func validCACert(cert, key []byte) bool {
+	valid, err := validCert(cert, cert, key, caName, lookaheadTime())
+	if err != nil {
+		return false
+	}
+	return valid
+}
+
+func validCert(caCert, cert, key []byte, dnsName string, at time.Time) (bool, error) {
+	if len(caCert) == 0 || len(cert) == 0 || len(key) == 0 {
+		return false, errors.New("empty cert")
+	}
+
+	pool := x509.NewCertPool()
+	caDer, _ := pem.Decode(caCert)
+	if caDer == nil {
+		return false, errors.New("bad CA cert")
+	}
+	cac, err := x509.ParseCertificate(caDer.Bytes)
+	if err != nil {
+		return false, errors.Wrap(err, "parsing CA cert")
+	}
+	pool.AddCert(cac)
+
+	_, err = tls.X509KeyPair(cert, key)
+	if err != nil {
+		return false, errors.Wrap(err, "building key pair")
+	}
+
+	b, _ := pem.Decode(cert)
+	if b == nil {
+		return false, errors.New("bad private key")
+	}
+
+	crt, err := x509.ParseCertificate(b.Bytes)
+	if err != nil {
+		return false, errors.Wrap(err, "parsing cert")
+	}
+	_, err = crt.Verify(x509.VerifyOptions{
+		DNSName:     dnsName,
+		Roots:       pool,
+		CurrentTime: at,
+	})
+	if err != nil {
+		return false, errors.Wrap(err, "verifying cert")
+	}
+	return true, nil
+}
+
+// controller code for making sure the CA cert on the
+// validatingwebhookconfiguration doesn't get clobbered
+
+var _ handler.Mapper = &mapper{}
+
+type mapper struct {}
+
+func (m *mapper) Map(object handler.MapObject) []reconcile.Request {
+	if object.Meta.GetNamespace() != vwhKey.Namespace {
+		return nil
+	}
+	if object.Meta.GetName() != vwhKey.Name {
+		return nil
+	}
+	return []reconcile.Request{{NamespacedName: secretKey}}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func addController(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(("validating-webhook-controller"), mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to the provided ValidatingWebhookConfiguration
+	s := &corev1.Secret{}
+	if err := c.Watch(&source.Kind{Type: s}, &handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
+
+	vwh := &unstructured.Unstructured{}
+	vwh.SetGroupVersionKind(vwhGVK)
+	mapper := &handler.EnqueueRequestsFromMapFunc{ToRequests: &mapper{}}
+	if err := c.Watch(&source.Kind{Type: vwh}, mapper); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileVWH{}
+
+// ReconcileVWH reconciles a validatingwebhookconfiguration, making sure it
+// has the appropriate CA cert
+type ReconcileVWH struct {
+	client.Client
+	scheme *runtime.Scheme
```

Comment:
Nit: This reconciler does not feel like an actual extension of k8s client or runtime scheme. Therefore I would name these variables rather than leaving them as anonymous extensions of the struct.

Reply:
Yeah, I'm not sure why Kubebuilder chose to make these extensions. Removing.